flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/sink/writer/FileWriter.java

```diff
@@ -0,0 +1,278 @@
+
+		this.activeBuckets = new HashMap<>();
+		this.bucketerContext = new BucketerContext();
+
+		this.bucketStateSerializer = new FileWriterBucketStateSerializer<>(
+				bucketWriter.getProperties().getInProgressFileRecoverableSerializer(),
+				bucketAssigner.getSerializer());
+	}
+
+	/**
+	 * Initializes the state after recovery from a failure.
+	 *
+	 * <p>During this process:
+	 * <ol>
+	 *     <li>we set the initial value for part counter to the maximum value used before across all tasks and buckets.
+	 *     This guarantees that we do not overwrite valid data,</li>
+	 *     <li>we commit any pending files for previous checkpoints (previous to the last successful one from which we restore),</li>
+	 *     <li>we resume writing to the previous in-progress file of each bucket, and</li>
+	 *     <li>if we receive multiple states for the same bucket, we merge them.</li>
+	 * </ol>
+	 *
+	 * @param bucketStates the state holding recovered state about active buckets.
+	 *
+	 * @throws Exception if anything goes wrong during retrieving the state or restoring/committing of any
+	 * 		in-progress/pending part files
+	 */
+	public void initializeState(List<FileWriterBucketState<BucketID>> bucketStates) throws IOException {
+		for (FileWriterBucketState<BucketID> state : bucketStates) {
+			BucketID bucketId = state.getBucketId();
+
+			if (LOG.isDebugEnabled()) {
+				LOG.debug("Restoring: {}", state);
+			}
+
+			FileWriterBucket<IN, BucketID> restoredBucket = bucketFactory.restoreBucket(
+					bucketWriter,
+					rollingPolicy,
+					state,
+					outputFileConfig);
+
+			updateActiveBucketId(bucketId, restoredBucket);
+		}
+	}
+
+	private void updateActiveBucketId(
+			BucketID bucketId,
+			FileWriterBucket<IN, BucketID> restoredBucket) throws IOException {
+		final FileWriterBucket<IN, BucketID> bucket = activeBuckets.get(bucketId);
+		if (bucket != null) {
+			bucket.merge(restoredBucket);
+		} else {
+			activeBuckets.put(bucketId, restoredBucket);
+		}
+	}
+
+	@Override
+	public void write(IN element, Context context) throws IOException {
+		// setting the values in the bucketer context
+		bucketerContext.update(
+				context.timestamp(),
+				context.currentWatermark());
+
+		final BucketID bucketId = bucketAssigner.getBucketId(element, bucketerContext);
+		final FileWriterBucket<IN, BucketID> bucket = getOrCreateBucketForBucketId(bucketId);
+		bucket.write(element);
+	}
+
+	@Override
+	public List<FileSinkCommittable> prepareCommit(boolean flush) throws IOException {
+		List<FileSinkCommittable> committables = new ArrayList<>();
+
+		// Every time before we prepare commit, we first check and remove the inactive
+		// buckets. Checking the activeness right before pre-committing avoid re-creating
+		// the bucket every time if the bucket use OnCheckpointingRollingPolicy.
+		Iterator<Map.Entry<BucketID, FileWriterBucket<IN, BucketID>>> activeBucketIt =
+				activeBuckets.entrySet().iterator();
+		while (activeBucketIt.hasNext()) {
```

Comment:
Shouldn't we check if a bucket is active _after_ we send the committables? A bucket may only have pending files, right? So after sending them, it may be empty and we can remove it. WDYT?

Reply:
One concern for the bucket that only have pending files is that we check it after sending the commits, it would always be inactive, then we will need to remove it and add a new one in the next checkpoint period.Do you think this would be a problem ?   If we delay the checking for one checkpoint period, we would only remove the bucket if it does not receive records for one checkpoint period.

Reply:
The only problem I can find is memory consumption. Nothing more. You think that this is not an issue?

Reply:
No, not this issue, I referred to the issue that we might need to recreate the bucket on each checkpoint period if `OnCheckpointRollingPolicy` is used and we detecting the inactiveness after send the committables. In this case the bucket will always be detected to be inactive since all committables are sent out right before the detection.